### ISSUE TITLE
Game stuck at "Descending..." after level 5

### ISSUE THREAD

**Issue by Kishan-Gupta** [00-Evan/shattered-pixel-dungeon#136]
SPD Version: 0.4.3a (from F-Droid)
Android version: 6.0.1

The game gets stuck at the "Descending..." screen when I try to get down from level 5. I've tried quitting and reloading over a dozen times but no luck so far.

Screenshot -

![sdp_descending_bug](https://cloud.githubusercontent.com/assets/8131771/19831442/c87da138-9e27-11e6-9f44-be8874b8a5bf.png)

**Comment by 00-Evan** [comment]
I made some changes to how the RNG of levelgen works in 0.4.3 and it seems to have provoked some existing low-probability levelgen faults (which is good, makes them easier to fix).

I implemented some tracking code into the google play version of shattered (it's opt-in, and not present in the f-droid build at all) to help with these levelgen issues in 0.4.3a, as a result I am aware of the cause of this one and hope to have it fixed in 0.4.3b.

**Comment by Kishan-Gupta** [comment]
Thanks for the quick reply.

Would you mind describing the cause of this bug in more detail? Both to satisfy my curiosity and so I could try and fix it myself for now - I have a bad case of SPD addiction ;)

**Comment by 00-Evan** [comment]
Sure, if you'd like to build the app yourself to fix the problem.

The offending code, (at least in the cases I have tested so far, it is possible there is more than 1 bug causing this), is in RegularLevel.Java lines 287 to 293:

```
while (count < 6) {
	Room r = randomRoom( Type.TUNNEL, 1 );
	if (r != null) {
		r.type = Type.STANDARD;
		count++;
	}
}
```

basically this is trying to ensure that there are at least 6 regular rooms (the variable count is initialized and incremented prior to this block of code). It loops and attempts to assign rooms until there are 6 regular ones. randomRoom( type, N ) picks a random room in the level, checks if it is of the corresponding type, and if it is it returns it. If it isn't it tries as many as N times, and then gives up and returns null. The potential for infinite looping here is obvious, but previously there were no (or at least very few) possible levels which created a situation where there were less than 6 possible standard rooms,

Internally I've changed it to this:

```
while (count < 6) {
	Room r = randomRoom( Type.TUNNEL, 20 );
	if (r != null) {
		r.type = Type.STANDARD;
		count++;
	} else
		return false;
}
```

We now try 20 times each iteration, and if we can't find anything then we give up and return false (which scraps the level and starts over).
